cli/worker.rs

```diff
@@ -349,55 +362,151 @@ impl CliMainWorker {
+    let mut tests = tests
+      .into_iter()
+      .filter(|(d, _)| filter.includes(&d.name))
+      .collect::<Vec<_>>();
+    if let Some(seed) = shuffle {
+      tests.shuffle(&mut StdRng::seed_from_u64(*seed));
+    }
+    sender.send(TestEvent::Plan(TestPlan {
+      origin: origin.to_string(),
+      total: tests.len(),
+      filtered_out: unfiltered - tests.len(),
+      used_only,
+    }))?;
+    let mut had_uncaught_error = false;
+    for (desc, function) in tests {
+      if fail_fast_tracker.should_stop() {
+        break;
+      }
+      if desc.ignore {
+        sender.send(TestEvent::Result(desc.id, TestResult::Ignored, 0))?;
+        continue;
+      }
+      if had_uncaught_error {
+        sender.send(TestEvent::Result(desc.id, TestResult::Cancelled, 0))?;
+        continue;
+      }
+      sender.send(TestEvent::Wait(desc.id))?;
+      let earlier = SystemTime::now();
+      let promise = {
+        let scope = &mut self.worker.js_runtime.handle_scope();
+        let cb = function.open(scope);
+        let this = v8::undefined(scope).into();
+        let promise = cb.call(scope, this, &[]).unwrap();
+        v8::Global::new(scope, promise)
+      };
+      let result = match self.worker.js_runtime.resolve_value(promise).await {
+        Ok(r) => r,
+        Err(error) => {
+          if error.is::<JsError>() {
+            sender.send(TestEvent::UncaughtError(
+              origin.to_string(),
+              Box::new(error.downcast::<JsError>().unwrap()),
+            ))?;
+            fail_fast_tracker.add_failure();
+            sender.send(TestEvent::Result(
+              desc.id,
+              TestResult::Cancelled,
+              0,
+            ))?;
+            had_uncaught_error = true;
+            continue;
+          } else {
+            return Err(error);
+          }
+        }
+      };
+      let scope = &mut self.worker.js_runtime.handle_scope();
+      let result = v8::Local::new(scope, result);
+      let result = serde_v8::from_v8::<TestResult>(scope, result)?;
+      if matches!(result, TestResult::Failed(_)) {
+        fail_fast_tracker.add_failure();
+      }
+      let elapsed = SystemTime::now().duration_since(earlier)?.as_millis();
+      sender.send(TestEvent::Result(desc.id, result, elapsed as u64))?;
+    }
     Ok(())
   }
 
   /// Run benches declared with `Deno.bench()`. Bench events will be dispatched
   /// by calling ops which are currently only implemented in the CLI crate.
-  pub async fn run_benchmarks(&mut self) -> Result<(), AnyError> {
-    let promise = {
-      let scope = &mut self.worker.js_runtime.handle_scope();
-      let cb = self
-        .js_run_benchmarks_callback
-        .as_ref()
-        .unwrap()
-        .open(scope);
-      let this = v8::undefined(scope).into();
-      let promise = cb.call(scope, this, &[]).unwrap();
-      v8::Global::new(scope, promise)
+  pub async fn run_benchmarks(
+    &mut self,
+    filter: TestFilter,
+  ) -> Result<(), AnyError> {
```

Comment:
I feel like this method and `run_tests` are so specific that they should be moved to relevant `cli/tools/` modules instead of residing in `cli/worker.rs`. Thoughts?

Reply:
Yeah I wanted to but there are these `CliMainWorker::run_test_specifier()` and `CliMainWorker::run_lsp_test_specifier()` in between that made it look annoying to approach. I think I'll start by inlining those..